User can buy a subscription

DIzosimov opened the pull request:
[PT Link](https://www.pivotaltracker.com/story/show/169285660)

## PR Content
- Implements react stripe elements
- PaymentForm and CheckoutForm components
- Integration test for accessing payment form
- Integration test for actually paying
- Fixes list article component
- Adds axios interception module to send in credentials to backend.
- Adds stripe, and hides API key.

## What we have learned
- How to implement stripe in react
- How to find iframe elements in the DOM for cypress
- The importance of version upgrades on software (e.g. cypress) that may cause bugs
- How to create custom cypress commands (in support/commands.js)
- How to hide API Keys
- How to intercept axios call to send in headers (credentials) to backend.

## Screenshots
<img width="561" alt="Screenshot 2019-11-05 at 12 26 46" src="https://user-images.githubusercontent.com/53824850/68204306-f8d82380-ffc7-11e9-9eb2-10cca5b02748.png">

<hr>

<img width="526" alt="Screenshot 2019-11-05 at 12 27 03 (1)" src="https://user-images.githubusercontent.com/53824850/68204317-fece0480-ffc7-11e9-8f7f-1bdd891541de.png">

<hr>

<img width="899" alt="Screenshot 2019-11-05 at 15 29 38" src="https://user-images.githubusercontent.com/53824850/68216466-37c6a300-ffe1-11e9-878d-659683e352b3.png">

DIzosimov committed:
Adds react stripe elements for stripe, adds PaymentForm component.

DIzosimov committed:
Adds userCanAccessPaymentForm integration test (happy path).

DIzosimov committed:
Adds checkoutform component and adds paymentform component to app component.

DIzosimov committed:
Adds stripe script to index.html, adds stripe api key to PaymentForm component, adds to checkoutform component.

DIzosimov committed:
Disables chrome web security in order to read iframe elements in cypress, adds integration test (happy path) for user successfully paying and accessing paymentform.

DIzosimov committed:
Refactors code.

DIzosimov committed:
refactors code.

DIzosimov committed:
Adds sadpath integration test for checkoutForm.

DIzosimov committed:
Removes config from cy/plugins/index.

beccaburns committed:
Reverts to previous commit 1ccbf9e.

beccaburns committed:
Refactors code.

piavW committed:
Updates integration test with correct path for auth.

beccaburns committed:
Merge branch 'stripe_payment' of https://github.com/DavveDavve/classy_news_client into stripe_payment

piavW committed:
Corrects url paths for paymentform tests.

beccaburns committed:
Merge branch 'stripe_payment' of https://github.com/DavveDavve/classy_news_client into stripe_payment

piavW committed:
Refactors code to hide paymentform behind a button.

DIzosimov committed:
Solves merge conflicts with upstream

beccaburns committed:
Resolves merge conflicts.

beccaburns committed:
Creates test to hide payment form from visitor.

beccaburns committed:
Refactors code.

piavW committed:
Refactors code to only display one welcomeMessage upon successful signin, adds attributes role and email to reduxTokenAuth.

piavW committed:
Refactors test due to spelling error.

piavW committed:
Removes unnecessary console.log

piavW committed:
Removes empty code.

piavW committed:
Refactors code.

GergKllai1 commented on src/Components/CheckoutForm.jsx:
```diff
@@ -0,0 +1,101 @@
+      ? this.stripePayment(token.id) 
+      : this.setState({
+        message: this.state.message
+      })
```
What does this do?

piavW replied:
It sets the message to the error-message we get from the Stripe payment function (which we pass into the submit payment function)

GergKllai1 replied:
Currently this function sets itself to itself, not the message from stripe

piavW replied:
We've fixed it

GergKllai1 commented on src/Components/PaymentForm.jsx:
```diff
@@ -0,0 +1,57 @@
+  renderPaymentForm = () => {
+    this.setState({
+      renderPaymentForm: !this.state.renderPaymentForm
+    })
```
This is a bad practice. React does setState asynchronously. It does when it has free resources. Pointing to `this.state` could lead to unintended end results. You should use prevState:

```
this.setState(prevState => ({renderPaymentForm: !prevState.renderPaymentForm )}) 
```

piavW replied:
modified

GergKllai1 commented on src/Components/Signup.jsx:
```diff
@@ -48,11 +49,10 @@ class Signup extends Component {
 
     if (this.props.currentUser.isSignedIn) {
-      welcomeMessage = <p id="welcome-message">Hello {this.props.currentUser.attributes.name}</p>
+          signupForm = ( <PaymentForm /> )
```
This is not a great practice, you should use 2 separate variables and toggle them conditionally instead.
It can be difficult for the future programmers to read your code if `signupForm` can be the signup and payment form as well.

piavW replied:
okay, we're planning on changing this when we do the navbar (which is the next feature)

piavW committed:
Creates getCredentials file, adds headers with credentials to relevant api request files. Refactors code.

piavW committed:
Changes response from API & hides stripe api key
- updates fixture files to match new response
- refactors components with new response code
- hides Stripe API key in env.development

piavW committed:
Refactors signup component code.

piavW committed:
Removes unnecessary empty lines.

miyeaier committed:
Remove redundant error message state.

oliverochman commented on cypress/integration/userCanAccessPaymentForm.spec.js:
```diff
@@ -0,0 +1,51 @@
+describe('Subscribe button is only visible for signed in user', () => {
+  beforeEach(() => {
+    cy.server()
+    cy.route({
```
why do you stub this request out?

DIzosimov replied:
We stub out the articles, since otherwise it will complain about the .map of undefined, but all we do in this test is check so a non-signed in user can't see the payment form

oliverochman replied:
ok by not in the beforeEach block then?

oliverochman commented on cypress/integration/userCanAccessPaymentForm.spec.js:
```diff
@@ -0,0 +1,51 @@
+  beforeEach(() => {
+    cy.server()
+    cy.route({
+      method: 'POST',
```
why is it a POST request?

DIzosimov replied:
changed :)

oliverochman commented on cypress/integration/userCanAccessPaymentForm.spec.js:
```diff
@@ -0,0 +1,51 @@
+  })
+})
+
+describe('Subscribe button is only visible for signed in user', () => {
```
why is this not a it block?

DIzosimov replied:
fixed :)

oliverochman commented on cypress/integration/userCanAccessPaymentForm.spec.js:
```diff
@@ -0,0 +1,51 @@
+    cy.get('#submit-signup-form').click()
+    cy.get('#paymentform-button').click()
+    cy.get('#subscribe-button').click()
+    cy.route({
```
why do you not stub this request out in the before block

oliverochman replied:
and why are you stubbing out successful_subscription_payment if you are not actually paying the subscription in this test

DIzosimov replied:
Very valid questions, and I'm stupid and reading through the test I'm not really sure wtf I was doing so I removed bunch of unecessary stuff and cleaned it up :)

oliverochman commented on cypress/integration/userCanPaySubscriptions.spec.js:
```diff
@@ -0,0 +1,107 @@
+describe('User can pay for subscription', () => {
```
why is `userCanAccessPaymentForm.spec.js` necessary if you have this test?

DIzosimov replied:
Mainly for the sadpath that if they are not signed in they can't access it.

oliverochman replied:
Then add the sad path to this file?

DIzosimov replied:
That's actually a good question, its pretty redundant right? :P

DIzosimov replied:
We removed the userCanAccessPaymentForm spec file and added the sadpath as extra it block to paysubscription test instead.

oliverochman commented on src/Components/Signup.jsx:
```diff
@@ -17,6 +18,7 @@ class Signup extends Component {
     name: '',
     city: '',
     country: 'Sweden',
+    role: 'user',
```
why do you set the role in state if you cant change it?

DIzosimov replied:
changed it to empty string :)

DIzosimov committed:
Refactors test and code.

DIzosimov committed:
Solves some merge conflicts in viewArticle component by sending in headers in module.

DIzosimov committed:
Refactors code to fix bugs with merge conflicts.

oliverochman commented on src/Components/Signup.jsx:
```diff
@@ -17,6 +18,7 @@ class Signup extends Component {
     name: '',
     city: '',
     country: 'Sweden',
+    role: '',
```
where do u set the role?

DIzosimov replied:
```
handleSignup = () => {
    const { registerUser } = this.props;
    const { email, name, nickname, password, password_confirmation, city, country, role } = this.state;
    registerUser({ email, name, nickname, password, password_confirmation, city, country, role })
      .then(
        console.log('yiihaaaa')
      )
      .catch(error => {
        this.setState({errorMessage: error.response.data.errors.full_messages[0]}) 
      })
  }
```

DIzosimov replied:
:)

oliverochman replied:
the role is empty in the constructor, where do u set the role to 'user'?

DIzosimov replied:
well, we pass in the props from registerUser attributes in reduxTokenAuthConfig and set the state of the user to whatever we have the user set to in the backend, however, maybe we dont need it in the signup and only in login?

DIzosimov committed:
Removes redundant test.

DIzosimov committed:
Hardcodes role to user for signUp component.
